llama-index-core/llama_index/core/memory/vector_memory.py

```diff
@@ -0,0 +1,213 @@
+"""Vector memory.
+
+Memory backed by a vector database.
+
+"""
+
+import uuid
+from typing import Any, Dict, List, Optional
+from llama_index.core.bridge.pydantic import validator
+
+from llama_index.core.schema import TextNode
+from llama_index.core.vector_stores.types import VectorStore
+from llama_index.core.base.llms.types import ChatMessage, MessageRole
+from llama_index.core.bridge.pydantic import Field
+from llama_index.core.memory.types import DEFAULT_CHAT_STORE_KEY, BaseMemory
+from llama_index.core.storage.chat_store import BaseChatStore, SimpleChatStore
+from llama_index.core.embeddings.utils import EmbedType
+
+
+DEFAULT_TOKEN_LIMIT_RATIO = 0.75
+DEFAULT_TOKEN_LIMIT = 3000
+
+
+def _stringify_obj(d: Any):
+    """Utility function to convert all keys in a dictionary to strings."""
+    if isinstance(d, list):
+        return [_stringify_obj(v) for v in d]
+    elif isinstance(d, Dict):
+        return {str(k): _stringify_obj(v) for k, v in d.items()}
+    else:
+        return str(d)
+
+
+def _stringify_chat_message(msg: ChatMessage) -> Dict:
+    """Utility function to convert chatmessage to serializable dict."""
+    msg_dict = msg.dict()
+    msg_dict["additional_kwargs"] = _stringify_obj(msg_dict["additional_kwargs"])
+    return msg_dict
+
+
+CUR_USER_MSG_KEY = "cur_user_msg"
+
+
+class VectorMemory(BaseMemory):
+    """Memory backed by a vector index."""
+
+    vector_index: Any
+    retriever_kwargs: Dict[str, Any] = Field(default_factory=dict)
+
+    # Whether to combine a user message with all subsequent messages
+    # until the next user message into a single message
+    # This is on by default, ensuring that we always fetch contiguous blocks of user/response pairs.
+    # Turning this off may lead to errors in the function calling API of the LLM.
+    # If this is on, then any message that's not a user message will be combined with the last user message
+    # in the vector store.
+    batch_by_user_message: bool = True
+
+    chat_store: BaseChatStore = Field(default_factory=SimpleChatStore)
+    # NOTE/TODO: we need this to store id's for the messages
+    # This is not needed once vector stores implement delete_all capabilities
+    chat_store_key: str = Field(default=DEFAULT_CHAT_STORE_KEY)
```

Comment:
yeah basically. i think we still need the cur_user_msg_key for batching, but we no longer need to have a chat store track _all_ user ids

Reply:
Sorry, I'm still hung up on this, but if we do a delete_all on the vector store, wouldn't that remove all past batches in addition to the current running batch? I think we would want to delete the current running batch so that we can start a new one with the incoming user message?

Reply:
left this for a future PR, but have started a working branch/PR for this that removes `chat_store`. I have effectively got it working, but it needs further polishing.

https://github.com/run-llama/llama_index/pull/13729